Purchases: Revamp Cancellation Flow

Aurorum opened the pull request:
Fixes #71577

## Proposed Changes

* Revamps and redesigns the Purchases cancellation flow 

## Why are these changes being made?
There are several issues with the current design: it's not particularly aesthetically pleasing; it can be confusing in terms of clearly displaying the refund amount; and I encountered a couple of bugs (eg. the subscription expiration date is highlighted in green when only the refund should be). As such, this screen was originally redesigned, but it got reverted because of the number of bugs introduced in that revamp.

This PR should bring back the design added in #70856, but hopefully without any of the bugs...

## Testing Instructions

There's a lot of different cases to test here, but in effect, the aim is to test the following cases:

- Navigating from Me > Purchases and from My Sites > Upgrades > Purchases (the styling is slightly different depending on the case)
- Subscription products vs one-off products (incl. Jetpack products) 
- Different plans - the WordPress.com plans will show a features list. Last time, there was a bug where the features list persisted on products which it wasn't applicable to, but that shouldn't happen here
- Domain attached to a plan or not attached to a plan 

I think screenshots are easier to demonstrate the different testing cases, so I'm going to attach a load of those instead. It's also worth noting that the testing instructions here are the same as #70856.



## WPcom Plans

- Should render a list of features which the user will lose once their plan ends
- Rather than featuring a box, which is a bit clunky given the list of features, should clearly include the refund in the text: `Once you confirm this cancellation, you will receive a full refund of £100.80 and your subscription will be removed immediately.`

| Before | After |
|--------|--------|
| <img width="1084" alt="Screenshot 2024-08-02 at 13 57 23" src="https://github.com/user-attachments/assets/b1c7abe6-6f71-47b9-9aed-4397eb19c6d8"> | <img width="1088" alt="Screenshot 2024-08-02 at 13 57 28" src="https://github.com/user-attachments/assets/b92d2fcc-e520-4ec4-905d-296047d5d155"> |

## Plan with domain

- Should render two boxes offering whether to keep the domain. Like on `trunk`, the user should confirm that understand that they might lose their domain (if they choose to cancel it) before the cancellation button becomes active. 
- If the domain is kept, the text should update to show the amount of a `partial refund` - not a `full refund`.  

| Before | After |
|--------|--------|
| <img width="1085" alt="Screenshot 2024-08-02 at 13 56 26" src="https://github.com/user-attachments/assets/84958539-47a2-45b6-8a6d-a1be1cee6d05"> | <img width="1084" alt="Screenshot 2024-08-02 at 13 56 33" src="https://github.com/user-attachments/assets/ef0dc5d0-913e-4f16-843d-98dede9c9041"> |

## Jetpack subscriptions

- Should just involve styling updates; no change to language here.

| Before | After |
|--------|--------|
| <img width="1091" alt="Screenshot 2024-08-02 at 13 59 25" src="https://github.com/user-attachments/assets/f1af83ef-9fb4-4971-829f-67493c01b377"> | <img width="1095" alt="Screenshot 2024-08-02 at 13 59 30" src="https://github.com/user-attachments/assets/d9c1d2bc-cc6b-405b-8107-710d2e746d10"> |

## Upgrades from My Sites > Upgrades > Purchases

- Verify that the styling here is all identical, especially if you're mostly testing this change against Me > Purchases. Although there shouldn't be any changes, Me > Purchases inherits some other styling and class names, which is what caused bugs here last time this was attempted. 
 
| Before | After |
|--------|--------|
| <img width="1087" alt="Screenshot 2024-08-02 at 13 59 45" src="https://github.com/user-attachments/assets/8e3b74c1-fd48-43b9-b932-9717df05ce9d"> | <img width="1102" alt="Screenshot 2024-08-02 at 13 59 40" src="https://github.com/user-attachments/assets/0cd4956f-6416-4fec-bb66-824f6c46bb5b"> |

## Other subscriptions

- No difference, except for the warning about subscriptions expiring. Whereas this was previously green because the code wrongly interpreted this as the refund amount, it should be red now. 

| Before | After |
|--------|--------|
| <img width="1079" alt="Screenshot 2024-08-02 at 13 59 58" src="https://github.com/user-attachments/assets/8ae26255-1ab3-430e-af4a-95ea143011fd"> | <img width="1094" alt="Screenshot 2024-08-02 at 14 00 02" src="https://github.com/user-attachments/assets/20a60215-c1bb-4072-932b-d11722d684dd"> |

## Standalone domains (ie. without a plan)

- Should be no difference except styling updates.

| Before | After |
|--------|--------|
| <img width="1073" alt="Screenshot 2024-08-02 at 14 00 13" src="https://github.com/user-attachments/assets/8113aca9-f1e0-47d9-8044-faa8e36a382a"> | <img width="1082" alt="Screenshot 2024-08-02 at 14 02 56" src="https://github.com/user-attachments/assets/f48b156c-5e6d-4e99-9d61-d191dbc1ffff"> |

# Styling changes for all states

- Should be two columns now, breaking to one column on mobile. 
- Confirm that the loading placeholder has been updated to reflect this.
- The Navigation menu has been removed in favour of a simple "Back" button, and some of the fonts/styles updated. 
- **The substance of the text and the cancellation button string should stay the same.** I've updated the capitalisation to be consistent though. 
- In effect, just make sure nothing looks off! 

cc @DavidRothstein, @michaeldcain, @southp - sorry, I'm sure there's a lot to review here, but I hope it's sort of the design which you had in mind (happy to make any changes!)

Aurorum committed:
Initial commit

Aurorum committed:
Revamp cancellation flow

jjchrisdiehl commented on client/me/purchases/cancel-purchase/button.jsx:
```diff
@@ -282,23 +282,23 @@ class CancelPurchaseButton extends Component {
 
 			if ( isDomainRegistration( purchase ) ) {
-				text = translate( 'Cancel Domain and Refund' );
+				text = translate( 'Cancel domain and refund' );
```
Minor nit, but it would be beneficial to convert this code into an IIFE similar to my comment [here](https://github.com/Automattic/wp-calypso/pull/89249#discussion_r1555047917). Granted, I know you didn't write the original code, but its a good place to update the existing codebase.

Aurorum replied:
Sure thing, done!

jjchrisdiehl reviewed (changes requested):
Hey @Aurorum, thanks for taking on this task! 😄

Could you rewrite the test instructions to detail the specific changes that need to be tested? Please specify any elements or stylings that have been corrected and what the reviewer should look for when reviewing.

It would be helpful to pair the written instructions with each individual before/after screenshot to make the review process clearer.

Thanks!

Aurorum committed:
Address review

Aurorum commented:
Hi @jjchrisdiehl - very fair shout, and that's a good idea with the testing instructions to break down each screenshot! I've updated the PR description now. I also realise you might know this flow a little better, so if I've missed anything (eg. a different type of purchase) that also seems broken when testing, please do let me know!

Aurorum committed:
Fix console errors

Aurorum committed:
Improve button alignment

Aurorum committed:
Remove redundant styles

JessBoctor commented:
Hi @Aurorum! Overall, this is looking great. I have a question about the styling changes for plans being cancelled: 

<img width="1016" alt="Screenshot 2024-08-08 at 15 47 38" src="https://github.com/user-attachments/assets/480173ac-b5ad-4e23-bfd9-fd436c9b8040">

`immediately` has been changed from normal font weight to bold. I appreciate that we want to convey importance to the customer having their products removed as soon as they cancel. However, I am worried this may cause unneeded anxiety about refunds. Refunds are also issued immediately, but it takes 5-7 business days for the cash to hit the customer's bank account. I am concerned this change might translate into customer's thinking that their cash will be in their account immediately too and then panic when it isn't there. 

<img width="767" alt="Screenshot 2024-08-08 at 15 51 53" src="https://github.com/user-attachments/assets/5e80c995-8f84-455f-a79a-24aab7dbfc29">

I think the same text, without the bold, is less panic inducing.

Aurorum committed:
Remove bold for "immediately"

Aurorum commented:
Thanks for taking a look @JessBoctor - that makes total sense! I've removed the bold for the word 'immediately'. :)

JessBoctor commented:
Hey @Aurorum! Let me say again, thanks so much for your efforts to improve the Cancellation flow. And, thanks for your patience. We did another review as a team, and found a few items we think need to be addressed. Here are some small things to start:

<img width="1075" alt="PR93211-08202024-Feedback-1through3" src="https://github.com/user-attachments/assets/a61489ab-bea1-4208-915b-d8d8b8f07625">

- (1) `Cancelling` vs `Canceling` → We should use `Canceling` (with a single `l`) for consistency, since it is already spelled that way in other places.
- (2) There is a missing paragraph break when canceling a plan which has a domain that is not being canceled
- (3) `Partial refund` vs `full refund` → Let's just remove the `partial` or `full` and state they will receive refund. This will cause the least amount of confusion and remove any chances for mislabeling. For example, in the `Domain with Plan` cancellation flow, the partial and full refund notation are backwards

![PR93211-full-partial-logic](https://github.com/user-attachments/assets/3782a148-ee3c-45d7-a4dc-975d301cca07)

There is some strange behavior happening when a plan and domain are both outside of the refund window, and only the plan is being cancelled: 

<img width="1078" alt="PR93211-Feedback-4through6" src="https://github.com/user-attachments/assets/6b566e8c-4db4-45b6-9330-d7970b0f33ff">

- (4) The notice that the customer's domain _won't_ be removed is shown twice
- (5) We don't need to show a calculated refund amount in this case, since there is not a refund being given
- (6) In the case of no-refund/products being removed at expiration, would it be possible to swap `Once you confirm this cancellation` for `If you complete this cancellation`? Since the subscription removal date is usually in the future, it sounds cleaner. This would be an improvement in all flows for a cancellation without a refund.


Thanks again and feel free to reach out if you have any questions!

Aurorum commented:
Thanks @JessBoctor - those are some very good spots! If it would be okay, I'm afraid that I'm going to have to ask for another sandbox cookie again though (cc @DavidRothstein). I didn't get round to using last week's one but it looks like it has expired now, so I can't test the purchase flow currently.

> There is some strange behavior happening when a plan and domain are both outside of the refund window, and only the plan is being cancelled

Is it also possible to add a purchase onto my account that's outside the refund window so I can test this? The problem with the sandbox is that by the time my purchase is outside the window, it gets wiped because it expires. :)

Thank you, and apologies for the inconvenience of going through this again!

Aurorum committed:
Address review

Aurorum commented:
Thanks so much @JessBoctor - that works perfectly! I didn't realise the sandbox purchases were saved but that makes it much easier. And thank you for the detailed review too! 

> (1) Cancelling vs Canceling → We should use Canceling (with a single l) for consistency, since it is already spelled that way in other places.

Fixed - sorry, joys of growing up on British English! :)

> (2) There is a missing paragraph break when canceling a plan which has a domain that is not being canceled

Fixed!

> (3) Partial refund vs full refund → Let's just remove the partial or full and state they will receive refund.

That makes sense, I've switched it to just 'refund'.

> (4) The notice that the customer's domain won't be removed is shown twice

Whoops, I completely missed that there were other domain states in that file. I've transferred them over and removed them from the top, so this should also be fixed now. A heads up that this also means that I've moved the message for cancelling domain mapping to beneath the features list too (I missed it in the `refund-information` file to start with!). 

> (5) We don't need to show a calculated refund amount in this case, since there is not a refund being given

This should be fixed, but it's the only one which I'm not certain about. From your top screenshot, it looks like that message should only appear when the subscription is refundable and the domain is not refundable. In other words, I believe that we have three states for domain registration:

- Plan not refundable/domain not refundable: `This plan includes the custom domain example.com, normally a £10 purchase. The domain will not be removed along with the plan, to avoid any interruptions for your visitors.` 
- Plan not refundable/domain refundable: `This plan includes the custom domain example.com, normally a £10 purchase. The domain will not be removed along with the plan, to avoid any interruptions for your visitors. You will receive a partial refund of £72 which is £82 for the plan minus £10 for the domain.`  
- Plan refundable/domain refundable: `What would you like to do with the domain?` and the two associated options.

The logic in the existing file seems pretty complicated, so please do let me know if I've understood that correctly. I've added that check for the line, which I'm pretty confident should work, but I don't think I can test that case without waiting a few days for a domain to expire.

> (6) In the case of no-refund/products being removed at expiration, would it be possible to swap Once you confirm this cancellation for If you complete this cancellation? Since the subscription removal date is usually in the future, it sounds cleaner. This would be an improvement in all flows for a cancellation without a refund.

Much cleaner - fixed! 

Thanks again, and please do let me know if there's still any issues! :)

sirbrillig commented on client/me/purchases/cancel-purchase/button.jsx:
```diff
@@ -275,42 +275,55 @@ class CancelPurchaseButton extends Component {
-		if ( hasAmountAvailableToRefund( purchase ) ) {
-			onClick = this.handleCancelPurchaseClick;
+		const onClick = ( () => {
+			if ( hasAmountAvailableToRefund( purchase ) ) {
```
A nice use of an IIFE!

These conditions are a little hard to grok, especially the else block (see [my notes on the issues of using else](https://payton.codes/2019/01/22/alternatives-to-else/)). Can we move all these nested conditions into a single level of indentation?

So, something like this:

```js
const isValidForRefund = isDomainRegistration( purchase ) || isSubscription( purchase ) || isOneTimePurchase( purchase );
if ( hasAmountAvailableToRefund( purchase ) && isValidForRefund ) {
 return this.handleCancelPurchaseClick;
}
const isValidForCancel = isDomainRegistration( purchase ) && isRefundable( purchase );
if ( ! hasAmountAvailableToRefund( purchase ) && isValidForCancel ) {
 return this.handleCancelPurchaseClick;
}
if ( ! hasAmountAvailableToRefund( purchase ) && isSubscription( purchase ) ) {
  return this.handleCancelPurchaseClick;
}
return () => this.cancelPurchase( purchase );
```

That way it's very explicit what each block represents. There's other ways to structure the above but hopefully you can get the idea from my example.

Aurorum replied:
Makes sense, thank you! I've made those changes now based on your example.

sirbrillig commented on client/me/purchases/cancel-purchase/button.jsx:
```diff
@@ -275,42 +275,55 @@ class CancelPurchaseButton extends Component {
-			onClick = () => {
-				this.cancelPurchase( purchase );
-			};
+		const text = ( () => {
```
Same comment here. Nested `if` inside `else` can get unwieldy quick. Can this be unwound?

Aurorum replied:
Done!

sirbrillig commented on client/me/purchases/cancel-purchase/domain-options.jsx:
```diff
@@ -0,0 +1,294 @@
+import FormRadio from 'calypso/components/forms/form-radio';
+import { getName, isRefundable, isSubscription } from 'calypso/lib/purchases';
+
+const CancelPurchaseDomainOptions = ( {
```
Ideally we are trying to avoid adding new files without TypeScript protections. I see that this file is just being extracted from another file that already is JS and I understand if you might not feel as confident converting things to TypeScript, in which case you could add `// @ts-check` as the first line in this file to see any potential issues that might be here. You can define any missing types [using JSDoc](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html). (Also true for the other JS files added by this PR.) 

Adding the types is not a requirement for this PR, however. We can always come in and add them afterward if needed, but it's something that's good to keep in mind. There's never an ideal time to convert things to TS but refactors are certainly not the worst time. Sometimes it can even expose bugs that already exist and have been hidden.

Aurorum replied:
Yes, I'm afraid that I'm not really familiar with TS at all. I've tried adding the `ts-check` on the file, but I didn't see any issues, which makes me think that I might be doing something wrong? As such, I hope that it's okay if this is out-of-scope for this PR now (though I appreciate that a refactor is a sensible time to include these sort of things).

sirbrillig replied:
No worries. We can make it a follow-up task.

sirbrillig commented on client/me/purchases/cancel-purchase/domain-options.jsx:
```diff
@@ -0,0 +1,294 @@
+	// Domain mapping.
+	if ( isDomainMapping( includedDomainPurchase ) ) {
+		if ( ! isRefundable( purchase ) ) {
+			return (
```
A minor suggestion: make each of these blocks into their own functional component (it can be inside this same file). That way the following code will be reduced from 236 lines down to something like the following (my names are probably not the best but this is the idea).

```js
	// Domain mapping.
	if ( isDomainMapping( includedDomainPurchase ) ) {
		if ( ! isRefundable( purchase ) ) {
			return <RefundableDomainMapMessage includedDomainPurchase={ includedDomainPurchase } />;
		}

		return <CancelableDomainMapMessage includedDomainPurchase={ includedDomainPurchase } />;
	}

	// Domain registration.
	if ( ! isRefundable( purchase ) ) {
		return <NonRefundableDomainPurchaseMessage includedDomainPurchase={ includedDomainPurchase } />;
	}

	if ( isRefundable( purchase ) && ! isRefundable( includedDomainPurchase ) ) {
		return (
			<RefundablePurchaseWithNonRefundableDomainMessage
				includedDomainPurchase={ includedDomainPurchase }
				planCostText={ planCostText }
				purchase={ purchase }
			/>
		);
	}

	return <DefaultRefundMessage />;
```

Hopefully it's apparent how much easier that is to read and follow.

Aurorum replied:
Sure thing, that does seem easier to follow! I've kept in the same function, but let me know whether that's what you had in mind. :)

sirbrillig commented on client/me/purchases/cancel-purchase/feature-list.jsx:
```diff
@@ -0,0 +1,52 @@
+import { getFeatureByKey } from '@automattic/calypso-products'; // eslint-disable-line import/named
```
Tell me more about `eslint-disable-line import/named` here. That error usually means that the function you're importing doesn't actually exist, which is dangerous to allow.

Aurorum replied:
I'm actually pretty stuck on this one - I'm not sure! `getFeatureByKey` definitely does exist, and the code (rightly) breaks when I exclude it.

https://github.com/Automattic/wp-calypso/blob/17999661e1882dd60527ed5a5753b236abe1e5bf/packages/calypso-products/src/main.ts#L943-L945

But in that case, I'm not sure why the Linter is throwing an error. I'd really appreciate some advice on if I'm missing something obvious!

sirbrillig replied:
I don't get any error in my local dev environment when I remove the comment. My guess is that what's happening is your local TypeScript cache has a version of `@automattic/calypso-products` from before `getFeatureByKey` was added. After you modify a package in the monorepo, sometimes you have to re-build the packages explicitly to update the cache. Usually I do this by `yarn build-packages`. Try that and let me know if you still see the error.

Aurorum replied:
Works! Thank you :)

sirbrillig commented on client/me/purchases/cancel-purchase/feature-list.jsx:
```diff
@@ -0,0 +1,52 @@
+				} ) }
+			</ul>
+			<p className="cancel-purchase__features-link">
+				<a href={ '/plans/my-plan/' + purchase?.domain }>
```
The null guard here is probably there for a good reason, but that means that if `purchase` isn't set, this will create a link to `/plans/my-plan/undefined` which is likely to be confusing for the user. Maybe we should bail earlier if there is no purchase? Either rendering nothing in this component at all or at least not rendering this link.

Aurorum replied:
Actually, I don't think there is good reason for a null guard and I've removed it. The features list is only rendered when there are available features compared against the current purchase, so if this is being rendered at all, it means that the purchase exists (unless I'm missing something!).

sirbrillig commented on client/me/purchases/cancel-purchase/index.jsx:
```diff
@@ -136,55 +141,132 @@ class CancelPurchase extends Component {
-				return this.props.translate(
-					'After you confirm this change, the domain will be removed immediately'
+				return translate(
+					'After you confirm this change, the domain will be removed immediately.'
```
These changes are fine but remember that they will require sending the strings to be retranslated so we should be sure to string freeze this PR and wait for the translations to be done before merging.

Aurorum replied:
Good shout!

sirbrillig commented on client/me/purchases/cancel-purchase/index.jsx:
```diff
@@ -218,68 +299,126 @@ class CancelPurchase extends Component {
+									<div className="cancel-purchase__footer-text">
+										{ hasAmountAvailableToRefund( purchase ) ? (
+											<p className="cancel-purchase__refund-amount">
+												{ this.renderFooterText( this.props ) }
```
`renderFooterText` doesn't accept any arguments. These sorts of errors are easy to spot if you add `// @ts-check` to the top of this file.

sirbrillig commented on client/me/purchases/cancel-purchase/index.jsx:
```diff
@@ -218,68 +299,126 @@ class CancelPurchase extends Component {
+											</p>
+										) : (
+											<p className="cancel-purchase__expiration-text">
+												{ this.renderExpirationText( this.props ) }
```
`renderExpirationText` also does not accept any arguments.

sirbrillig commented on client/me/purchases/cancel-purchase/index.jsx:
```diff
@@ -218,68 +299,126 @@ class CancelPurchase extends Component {
+							</>
+						) : (
+							<>
+								<p>{ this.renderFullText( this.props ) }</p>
```
`renderFullText` also does not accept any arguments.

sirbrillig commented on client/me/purchases/cancel-purchase/index.jsx:
```diff
@@ -218,68 +299,126 @@ class CancelPurchase extends Component {
 
 		if ( isSubscription( purchase ) ) {
-			heading = this.props.translate( 'Cancel Your %(purchaseName)s Subscription', {
+			heading = this.props.translate( 'Cancel your %(purchaseName)s subscription', {
```
Nice case fix here. These sneak in everywhere.

sirbrillig reviewed (commented):
This looks good! It's a big PR (dealing with an already-complex component) which makes it tough to review safely but I think @JessBoctor and @jjchrisdiehl have done a great job so far. I've added some of my own observations but they are mostly minor and will help a final review.

sirbrillig commented:
Of course! I was just back from sabbatical so I've had to learn how to do that but I've finally sent it along.

Aurorum committed:
Address feedback

Aurorum commented:
Thanks so much @sirbrillig! Have addressed your comments and made two other changes:

1) Included Odie within the support link based on @jjchrisdiehl's work in #94213 (I've just moved this to another file though because that one was becoming a little unwieldy with these changes!)
2) Made the layout full height again - I believe this was a regression introduced in #94180 and it already exists on production now, so it might be the case that these styles can be removed if it gets fixed earlier. 

Please let me know if there are any further issues!

sirbrillig reviewed (approved):
This looks good to me! I'll give it one more test soon and if it all works, I'll merge it. Thank you!

DavidRothstein commented:
Great work overall -- thank you!!!

I did a bit more testing now (following our original tests a few weeks ago) and noticed some inconsistencies -- I think these are not intentional?

1. When cancelling a domain, there is a "Back" link at the top of the initial screen:

![cancel-domain](https://github.com/user-attachments/assets/27556335-5307-4397-9a4b-7e06e7c93ab9)

But not on the confirmation screen that comes next:

![cancel-domain-confirm-after](https://github.com/user-attachments/assets/2f509f84-4507-46a9-bc1c-f1df46d00e6a)

Previously there was a back link on the confirmation screen also:

![cancel-domain-confirm-before](https://github.com/user-attachments/assets/b92b6ffe-a785-402c-b5b5-bc684520928a)

2. The above screenshots were taken when cancelling from the site-specific purchases page.  But when cancelling from the account-level purchases page (`/me/purchases`), I noticed the "Purchases" label at the top has gone missing -- compare this to the first screenshot above, and also to the current production behavior which has it:

![cancel-domain-me-purchases](https://github.com/user-attachments/assets/2cce13d5-d8d0-4adc-be35-8a59f67e42fc)

DavidRothstein commented:
Noticed a couple other things, for the case where you're cancelling a plan outside the refund window:

![cancel-plan-nonrefundable](https://github.com/user-attachments/assets/31c4b52b-56f7-437a-81e4-1e0ed6c51f9b)

1. The "By canceling..." language is not accurate, since in this case the plan won't be removed upon cancellation and therefore the features won't actually stop being available yet.  In the previous implementation of these new cancellation screens (the one that was eventually rolled back) the language used for this scenario was "When your WordPress.com Personal plan expires..." instead.  (The "By canceling..." language was reserved for the case where the plan is being removed right away, e.g. cancellation during the refund window.)
2. The "This plan includes the custom domain" paragraph is a bit misleading in this scenario (where the plan itself is not actually being removed).  By the time the plan is removed later, it's entirely possible that the domain actually will expire and be removed then also.  Would it make sense to remove the whole paragraph?  (Note: I see now that this is actually a preexisting bug since the same text appears in the current cancellation flow also -- although less prominently -- so we don't actually have to fix this now.)

Also, a minor issue (and this one is definitely not a blocker, but just wanted to note it anyway), I saw this when cancelling a domain outside the refund window:

![cancel-domain-after](https://github.com/user-attachments/assets/392c9a96-17e7-4057-8e70-5719d5060776)

On desktop, the line break before the "Cancel domain" button looks a bit odd; it would be better if the text were left aligned and the button right-aligned (with the text spilling over into another line in the left column) but again, not a deal-breaker.

Another minor issue, on the same page but at mobile screen widths, looks like there's a missing space after the question mark:

![cancel-domain-mobile](https://github.com/user-attachments/assets/ba984100-7e1b-404d-864e-dd3812c4df37)

Aurorum committed:
Address review

Aurorum commented:
Thanks @DavidRothstein!

> When cancelling a domain, there is a "Back" link at the top of the initial screen:

I must have missed this screen - fixed now!

> But when cancelling from the account-level purchases page (/me/purchases), I noticed the "Purchases" label at the top has gone missing -- compare this to the first screenshot above, and also to the current production behavior which has it:

I actually did this intentionally because the design in #70856 also removed the "Purchases" label. Personally, I think that the page works better without it - it doesn't really seem to fit now that the page is full-screen. I'm happy to change it if you'd prefer though! 

> The "By canceling..." language is not accurate, since in this case the plan won't be removed upon cancellation and therefore the features won't actually stop being available yet.

Ah, that makes sense! Should be fixed now. 

> The "This plan includes the custom domain" paragraph is a bit misleading in this scenario (where the plan itself is not actually being removed). 

I haven't made any changes here since, as you mentioned, this is the existing behaviour on production. If you'd like to remove the whole paragraph or rephrase it though, I'd be happy to make those changes! I just wasn't sure what would make most sense here and thought best to leave it for now without further direction. 

> On desktop, the line break before the "Cancel domain" button looks a bit odd; it would be better if the text were left aligned and the button right-aligned (with the text spilling over into another line in the left column) but again, not a deal-breaker.

Fixed!

> Another minor issue, on the same page but at mobile screen widths, looks like there's a missing space after the question mark:

Also fixed!

Thanks again for the review - please do let me know if there any further issues. :)

Aurorum commented:
Oh, also, one other change which I made while working on the Cancel Domain screen: I've wrapped the informational text in a warning card. I feel like this distinguishes the (dynamic) warning from the (fixed) text at the top of the screen and emphasises the content within it. As ever though, happy to change if this doesn't make sense! 

| Before | After |
|--------|--------|
| <img width="1057" alt="Screenshot 2024-09-10 at 21 02 05" src="https://github.com/user-attachments/assets/6fea9e5f-75c2-441d-95e1-5e7def61927e"> | <img width="1145" alt="Screenshot 2024-09-10 at 21 02 00" src="https://github.com/user-attachments/assets/e721f19f-365c-4e9b-b1cf-0a55ab709399"> |

sirbrillig commented:
> I haven't made any changes here since, as you mentioned, this is the existing behaviour on production. If you'd like to remove the whole paragraph or rephrase it though, I'd be happy to make those changes! I just wasn't sure what would make most sense here and thought best to leave it for now without further direction.

IMO these changes are large enough. We can take care of that paragraph in a follow-up.

Aurorum committed:
Remove linter comment

jjchrisdiehl commented:
Since this PR has gotten rather large, I have attempted to summarize the testable feedback:

```
- Check both /me/ and /my-sites/ entry points to Purchases
    - Make sure the styling is identical, /me/ inherits styling and class names from the profile page whereas /my-sites/ inherits from Calypso
- Check subscriptions and one-off products (incl. Jetpack and Akismet)
- Check different [WordPress.com](http://wordpress.com/) plans, including monthly, annual etc
    - Should render list of features which user will lose once plan ends
    - Features are aligned correctly and link to view all is clickable
    - Refund amount is correct and doesn’t over-promise
    - Text is correct, no errant punctuation for titles
    - Mobile responsive
- 100 Year plan?
- Bundled plans vs non-bundled plans and domains
    - Renders two radio boxes offering to keep the domain with a separate charge or cancel it.
    - If domain is kept it should show the amount of a partial refund and not full refund
- Jetpack will just be styling updates, no copy changes
- Check RTL support
- Check languages with known longer words
- Other subscriptions, no real change other than the green text is now red since it is apparently not a refund notice.
- Styling changes
    - Two columns now
    - Loading placeholder updated to reflect this
    - Nav menu removed in favour of back button
    - Fonts and styles updated
    - Updated some capitalization for consistency
    - Make sure ‘immediately’ is not bold
- Team review changes (see Jess’ comment on PR)
    - Use canceling instead of cancelling
    - Missing para break when canceling a plan that has a domain which will not be canceled
    - Remove ‘partial’ and ‘full’ from refund
    - Strange behaviour when plan and domain are both outside of the refund window and only the plan is being canceled. Basically duplication of the notice that the plan will not be removed.
    - No need to show calculated refund in this case since there is no refund to be given (both are outside their refund window)
    - In the case of no-refund/products being removed at expiration, change ‘Once you confirm this cancellation’ for ‘If you complete this cancellation’
- A couple PRs had to be accounted for https://github.com/Automattic/wp-calypso/pull/93211#issuecomment-2334034201
    - Ensure Odie ‘Ask a happiness engineer’ link still opens Chat/helpcenter
    - Make sure layout is full height again based on these changes
- David comments https://github.com/Automattic/wp-calypso/pull/93211#issuecomment-2341391926
    - When canceling a domain, there is a “back” link at the top of the initial screen but not on the confirmation screens that comes next
    - Previously there was a back link on the confirmation screen also (looks like a navigation component back link)
    - “Purchases” header at top of page is missing on /me/purchases, this is also observed in production
    - When canceling a plan outside the refund window, the ‘by canceling language’ is not accurate as it reflects a cancellation within the refund window.
        - Instead it should read "When your WordPress.com Personal plan expires..."
    - “This plan includes a custom domain” assumes the plan is being removed, but in this case the plan is just being canceled and will be removed at another time, possibly when the domain will also be removed. Suggests hiding this paragraph
    - The ‘After you confirm this change…’ line for the ‘Cancel domain’ button awkwardly sits above the Cancel domain button, suggests two columns where the text spills onto the next line
    - Lack of space between ‘Need help with your purchase” and “Ask a Happiness Engineer” - likely related to the Odie changes above
```

> I've wrapped the informational text in a warning card. I feel like this distinguishes the (dynamic) warning from the (fixed) text at the top of the screen and emphasises the content within it

I like this change, I wish we had some sort of blue `notice` prop instead, but in lieu of that the golden warning styling fits well enough here since it is attempting to warn about a destructive action.

jjchrisdiehl commented on client/me/purchases/cancel-purchase/style.scss:
```diff
@@ -1,96 +1,195 @@
-.cancel-purchase__support-information {
-	font-size: $font-body-small;
+.cancel-purchase__features-link {
+	text-align: right;
```
Might make more sense to left or center align this. When it is aligned right it feels like it is floating out of alignment compared to the featured list items.

If you use flexbox here you can more easily account for RTL layouts too:

```
.cancel-purchase__features-link {
    display: flex;
    justify-content: flex-start; // or center
    margin-top: 1em;
}
```

Aurorum replied:
Makes sense! I'm curious as to whether you'd recommend centre-aligning or left-aligning this? 

**Centre:**
<img width="1059" alt="Screenshot 2024-09-11 at 21 54 37" src="https://github.com/user-attachments/assets/15d478c6-5603-4a71-b831-7c4c4039f662">

**Left:**
<img width="1062" alt="Screenshot 2024-09-11 at 21 55 28" src="https://github.com/user-attachments/assets/0bd4c51d-5358-47be-b4b5-9b6c33cdd5e9">

Just an idea, but one thing I feel with both the above options is that the text feels a bit crowded with the link. To avoid this, I quite like the idea of treating the features list almost like its own card, but that might just be my preference.  

<img width="1063" alt="Screenshot 2024-09-11 at 21 54 04" src="https://github.com/user-attachments/assets/891624e2-afcf-4b2f-9a1c-b07f863017f1">

Let me know what you think (and thank you for going through and summarising all the feedback!). :)

jjchrisdiehl replied:
I agree that providing an outline to the features list section does help distinguish it from the rest of the content. Also, at that point, left, center, right doesn't really matter since the border provides a better anchor point for the link.

I am okay with moving forward with this idea, my personal preference is to center the 'View all features' link within the card.

jjchrisdiehl reviewed (changes requested):
### Which Calypso?
I think we might be blending two different 'Calypso' styling paradigms in this PR. 

As you can see, we used to use a 'light gray background, white card with border, separate headercake' layout which I'll unofficially refer to as 'Calypso dashboard v1' styling. This can be seen in the 'Purchases' and 'Emails' sections, to name a few.

Whereas, over the past two years, portions of the Calypso dashboard has changed over to a minimalist, white background, no borders, 'v2' styling. You can see this in places like 'My Home', 'Stats', 'Plans', etc.

What I see happening here is that the different sections of Purchases are still using v1 whereas the [original PR](https://github.com/Automattic/wp-calypso/pull/70856) attempted to update the cancellation section to v2, among other changes.

I'm not sure what the decision making process was for just targeting the cancelation page, but I believe a wider effort to:

1. Confirm that v2 is still the direction Calypso design is moving in 
2. And, if so, refactor the remaining Purchases sections to match v2

Pinging @sirbrillig @DavidRothstein and maybe even @ianstewart for thoughts on how to handle Calypso styling here.

This probably deserves its own issue anyway. Added one here https://github.com/Automattic/payments-shilling/issues/3112

### How to handle this
With that being said, I think the changes in this PR should try to preserve the v1 styling and focus primarily on the structural and functional changes found in the card itself.

For instance, I think rolling back the header to the separated HeaderCake found in v1 while also including the v2 styles and structure you added within the card would strike a good balance. 

Then, at a later point when we decide to update the Purchases styling to v2, we'll be in a good place to handle this.

### Bug
You'll see that in the /me/purchases version of the changes we're missing the Purchases title as seen in production:

| Pull Request | Production |
| ----- | -----|
| ![image](https://github.com/user-attachments/assets/d72c1a10-767a-4fa3-9685-878618a4db98) | ![image](https://github.com/user-attachments/assets/9618e029-cbdb-49ed-9b01-9fc455a44ae1) |


Let me know if you have any questions!

sirbrillig commented:
> You'll see that in the /me/purchases version of the changes we're missing the Purchases title as seen in production:

@jjchrisdiehl  I think that was by design. See the reply here: https://github.com/Automattic/wp-calypso/pull/93211#issuecomment-2341908450

jjchrisdiehl commented:
> @jjchrisdiehl I think that was by design. See the reply here: [#93211 (comment)](https://github.com/Automattic/wp-calypso/pull/93211#issuecomment-2341908450)

Thanks for the context. It looks like all of the other sections in /me/ use this header component for the section title. I'm a bit hesitant to introduce a different "layout" when we're sort of in the middle of reducing the number of styles and layouts throughout Calypso.

I checked another section that offers a similar "modular" feel - the Add Payment Method section - and it seems to retain the v2 header while using its own v1 card:

<img width="1726" alt="image" src="https://github.com/user-attachments/assets/1aa94e9f-9d1e-48ef-b273-052dab9fd29c">

If we were to update these sections in the future, it may be worthwhile to actually use the "header" as a breadcrumb (which is the component it appears to be using) and adjust things such as:
- Change the title to "Purchases > Add Payment Method"
- Remove the headercake for the back button and use the standalone back button
- Update design to no longer use, or use less, borders (this would coincide with changes to the other sections of /me/ to keep things similar)

Anyway, I'm not against dropping the 'Purchases' header from the 'Cancel' section, since it doesn't really affect the user experience that much, but we would want to note the divergence of the design in case we receive further instruction to update this styling to match any broader changes with Calypso.

### Side note

Interestingly, the production version of Cancel doesn't seem to take the full height of the page:

<img width="1728" alt="image" src="https://github.com/user-attachments/assets/ea82a370-0304-4589-94ba-6cef5f728d1b">

While this PR's version does - so nothing to do here, just wanted to say "good job"! 😄

Aurorum commented:
Thanks @jjchrisdiehl! I hear your point about the discrepancy between Calypso styling. Out of interest then, are you envisaging something like this?

<img width="1517" alt="Screenshot 2024-09-12 at 10 16 11" src="https://github.com/user-attachments/assets/5c12a634-dca6-42b1-9984-7e31fd083e1e">

Or perhaps something like this?

<img width="1508" alt="Screenshot 2024-09-12 at 10 21 19" src="https://github.com/user-attachments/assets/b5c76841-b9c1-4322-b5cf-b96984297abd">

Let me know what you have in mind and I'd be happy to change them. :)

sirbrillig commented:
I think that given the amount of changes, it's worth it to merge this PR as-is and deal with additional changes as a follow-up. I've added the String Freeze label to begin the translation process and we can merge it when that is complete.

jjchrisdiehl commented:
> I think that given the amount of changes, it's worth it to merge this PR as-is and deal with additional changes as a follow-up. I've added the String Freeze label to begin the translation process and we can merge it when that is complete.

I agree, the styling changes can be a follow up issue.

> Thanks @jjchrisdiehl! I hear your point about the discrepancy between Calypso styling. Out of interest then, are you envisaging something like this?

@Aurorum Personally, I prefer option two, since the borders create a better visual layout, whereas in option one it feels the two columns sort of blend together and become more visually confusing. We can experiment more with this in a new issue once this PR is merged.

Aurorum commented:
Sounds good - thanks both!

sirbrillig commented:
@Automattic/i18n I think there's something wrong with the translation process on this diff, presumably because it's an OSS contributor PR. Any ideas how I can get its strings translated?

yuliyan commented:
> @Automattic/i18n I think there's something wrong with the translation process on this diff, presumably because it's an OSS contributor PR. Any ideas how I can get its strings translated?

@sirbrillig, Since the branch doesn't exist in the `wp-calypso` repository, the translate build fails as it attempts to compare the branch with trunk to determine which lines have been changed and extract only the related translatable string updates..

The quickest solution I can think of would be to push a new branch with the changes from this one and open a new PR.

sirbrillig commented:
Moved to a new PR in the calypso repo here so we can get the translations working: https://github.com/Automattic/wp-calypso/pull/94864

Unfortunately I've also run into some problems. Despite what GitHub tells us here, this PR conflicts heavily with https://github.com/Automattic/wp-calypso/pull/94213. It also contains a merge commit (ugh, why does GitHub even let you do that) which makes fixing those conflicts very difficult. 😩  I've tried to resolve those issues in the new PR, but I'm not sure I got everything. We'll have to test everything added in this PR and in https://github.com/Automattic/wp-calypso/pull/94213 to make sure they both still work as intended.

Aurorum commented:
Thanks @sirbrillig! Ah, that's strange about the conflicts - I thought that I had resolved those and added the Odie support link to the implementation here, so not sure GitHub is doing here. Let me know if there's anything which I can do to help! :)

sirbrillig commented:
@Aurorum It's possible I messed up somehow during the rebase on my PR. Could you give it a try on your branch and rebase it onto a current version of trunk?

Aurorum commented:
Thanks @sirbrillig - done! :)

sirbrillig commented:
Thank you! However, it's best if we _rebase_ this branch, not create more merge commits. They're going to make it more complex.

There's a lot of documentation out there about rebasing, but since this a fork, I think there's two options:

To use the GitHub UI, I think you follow [these instructions](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/keeping-your-pull-request-in-sync-with-the-base-branch) and select "Update with rebase". 

Alternatively, [sync your fork to the main repo](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) and then `git rebase trunk`.

sirbrillig commented:
That looks much better. Thanks! I'm trying to duplicate the branch again to get translations working in https://github.com/Automattic/wp-calypso/pull/95065

sirbrillig commented:
This has been merged as https://github.com/Automattic/wp-calypso/pull/95065
